feat: implement configuration, context, typing, gql server

matthew-chirgwin opened the pull request:
- adds configuration types
- adds configuration utilities
- move configuration from server to config
- adds basic context for configuration/feature flags in client
- updates doc
- adds apollo server to express in config and mockapi modules
- starts to flesh out config server module
- starts to flesh out mockapi module
- adds tests
- fixes styling bug in Home component

Contributes to: #13
Contributes to: #50

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>
Signed-off-by: Nic Townsend <nictownsend@uk.ibm.com>

matthew-chirgwin committed:
feat: implement configuration, context, typing, gql server

- adds configuration types
- adds configuration utilities
- move configuration from server to config
- adds basic context for configuration/feature flags in client
- updates doc
- adds apollo server to express in config and mockapi modules
- starts to flesh out config server module
- starts to flesh out mockapi module
- adds tests
- fixes styling bug in Home component

Contributes to: #13
Contributes to: #50

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin commented:
There will be a following PR behind this to complete a few outstanding pieces of work:

- look into moving the server types to config
- templating key items into index html
- merging runtime overrides with the defined config
- properly itegrate the config context to call the config endpoint

This however is complete otherwise - enough that others can start to work with it. There is overlap between this issue and #50 - I have added a basic GQL server with test to the mockapi module - but this needs more work (which will be covered in #50)

matthew-chirgwin committed:
fix: remove gql file, define schema in code

- removes gql file, would need to be included in build. Will defer to
issue #50

Contributes to: #13
Contributes to: #50

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

JQrdan commented on client/Contexts/ConfigFeatureFlag/Context.tsx:
```diff
@@ -0,0 +1,42 @@
+  );
+};
+
+export { ConfigFeatureFlagProvider, ConfigFeatureFlagConsumer };
```
The React hooks docs says that the full context needs to be used for the useContext hook so can that be exported too. https://reactjs.org/docs/hooks-reference.html#usecontext

matthew-chirgwin replied:
> The React hooks docs says that the full context needs to be used for the useContext hook so can that be exported too. https://reactjs.org/docs/hooks-reference.html#usecontext

Indeed - In the next PR, I will have a `useConfigFeatureFlag` hook (which wraps `useContext`) so the context itself is private, but still accessible

JQrdan replied:
Ok perfect! As long as it's accessible through a hook

nictownsend commented on client/Contexts/ConfigFeatureFlag/ConfigFeatureFlag.spec.tsx:
```diff
@@ -0,0 +1,59 @@
+          </ConfigFeatureFlagConsumer>
+        </ConfigFeatureFlagProvider>
+      );
+      expect(getByText('{"example":{"value":true}}')).toBeInTheDocument();
```
I'd prefer if we seeded the test with the config, rather than test the inbuilt default? (I see later it's just dummy code but would be good). Unless of course the context should always be in charge of fetching and thus you don't want ability to override it?

Also, I know test docs state describe/it based "tdd" for `client/Contexts` - but wondering whether gherkin actually still applies here?

Or alternatively:

```
const setUp = (renderFn) => 
(<ConfigFeatureFlagProvider>
  <ConfigFeatureFlagConsumer>
     {renderFn}
  </CFFC>
</CFFP>
```

matthew-chirgwin replied:
This will be changing in the next PR, where with the introduction of the apollo pieces, seeding real 'mocked' response data was a requirement for the provider to be tested properly. I also have made some changes to the tests to use more helper code, akin to your `setUp` example in that PR

matthew-chirgwin replied:
https://github.com/matthew-chirgwin/strimzi-ui/tree/13-integrate/client/Contexts/ConfigFeatureFlag

nictownsend commented on client/Contexts/ConfigFeatureFlag/ConfigFeatureFlag.spec.tsx:
```diff
@@ -0,0 +1,59 @@
+        <ConfigFeatureFlagProvider>
+          <ConfigFeatureFlagConsumer>
+            {({ loading, error, isComplete, triggerRefetch }) => {
+              expect(loading).toBe(false);
```
Have you tried failure cases? Intrigued if expectation throwing error from render function ends the test properly

matthew-chirgwin replied:
The error here is placeholder - from the cases I have seen plumbing in an actual apollo request, no excpetion is thrown - we are instead provided with the error to do with whatever we want. I have tests for this in the next PR

nictownsend replied:
Apologies, I meant if the `expect` fails in the test, how the test runner handles it

matthew-chirgwin replied:
It handles it - and doesnt (seemingly) effect other tests:

```
 ● ConfigFeatureFlag › `ConfigFeatureFlagProvider` and `ConfigFeatureFlagConsumer` components › initial state › returns the expected state values on mount

    expect(received).toBe(expected) // Object.is equality

    Expected: false
    Received: true

      119 |                   rawResponse,
      120 |                 }) => {
    > 121 |                   expect(loading).toBe(false);
          |                                   ^
      122 |                   expect(error).toBe(false);
      123 |                   expect(isComplete).toBe(false);
      124 |                   expect(rawResponse).toEqual({});
```

nictownsend commented on client/Contexts/ConfigFeatureFlag/Context.tsx:
```diff
@@ -0,0 +1,42 @@
+  ...contextStates,
+});
+const ConfigFeatureFlagConsumer = ConfigFeatureFlag.Consumer;
+const ConfigFeatureFlagProvider: FunctionComponent = ({
```
I know this will be changed when you add in the data fetching. But I think it would be good to show an example in here of how the updating will work.

I'm hoping via a mixture of scoping, you will end up calling `triggerRefetch` -> query progress updates the `contextState`  -> updates config/featureFlags -> triggering the `useState` to re-render -> Provider to update

nictownsend replied:
Also, TS typing likely needed here - especially to ensure that the `defaultValue` to the context matches the shape that we later pass to `value` in provider

matthew-chirgwin replied:
> I know this will be changed when you add in the data fetching. But I think it would be good to show an example in here of how the updating will work.
> 
> I'm hoping via a mixture of scoping, you will end up calling `triggerRefetch` -> query progress updates the `contextState` -> updates config/featureFlags -> triggering the `useState` to re-render -> Provider to update

The flow will be the query will trigger on mount, and when a result (success or otherwise) is returned, a re render will occur for all children.

The refetch I was expecting would only be used in error scenarios, but there is nothing stopping it being used at any time. The result of the refecth will equally trigger a re render.

As for typing, I belive both here and in the next set of changes the typing is provided/correct. The next PR gets a bit more interesting due to the dynamic shape of what comes back, but the shape will follow what is defined in `config/types`

nictownsend commented on client/Contexts/ConfigFeatureFlag/Context.tsx:
```diff
@@ -0,0 +1,42 @@
+const featureFlags = { capability: { enabled: false } };
+
+// Loading/Error/Trigger Refetch etc
+const contextStates = {
```
Guessing this will be replaced by a `gql` query hook?

matthew-chirgwin replied:
It will indeed - placeholder in case anyone needed it

nictownsend commented on config/configHelpers.ts:
```diff
@@ -0,0 +1,75 @@
+} from './types';
+
+/** helper function to get the value of an environment variable, or a defined fallback value */
+export const getEnvvarValue: (
```
What is this sorcery? (might be overkill but https://www.logicbig.com/tutorials/misc/typescript/interface-describing-function.html) defining interface above and using it in the function?

matthew-chirgwin replied:
This is how I have been typing functions! :) A bit involved in this case with the curry, but my general approach is if the typing is not re used, I define it fully alongside/as a part of the function defintion as well. The interface would do the same thing, but in two stages - so if the signiture/type was shared, I would have made it one

pmuir replied:
I've always use the approach defined in the blog - personally I think it's possibly a bit more readable

nictownsend commented on config/configHelpers.ts:
```diff
@@ -0,0 +1,75 @@
+) => () => configurationLiteralValue = (
+  name,
+  fallback = `No value for ENVVAR ${name}`
+) => () => process.env[name] || fallback;
```
Probably fine - but `configurationLiteralValue` is a type union, but process.env will only ever return `string` - which (may) cause some confusion when using it with a fallback that isn't a string.

matthew-chirgwin replied:
Fair point - will update in the next PR

nictownsend commented on config/configHelpers.ts:
```diff
@@ -0,0 +1,75 @@
+) => {
+  // merge all provided configs
+  const mergedConfig = config.reduce(
+    (acc, thisConfig) => merge(acc, thisConfig),
```
Are there any of the following use cases expected:
1. Two configs have the same key and primitive value
2. Two configs have the same key with an array value


If so - merge will override value with the latest (in array case see : https://github.com/lodash/lodash/issues/2872)

matthew-chirgwin replied:
I expect we could have the same key defined in two files, with different values (primitive or otherwise). Eg a static default, and runtime 'real' perhaps. It means the order the configs are provided matter. Think a jsdoc comment would suffice?

nictownsend commented on config/configHelpers.ts:
```diff
@@ -0,0 +1,75 @@
+        valueType === 'number';
+      const isConfigurationValue =
+        !isLiteralValue &&
+        (value as programmaticValue<unknown, unknown>).configValue;
```
Won't this return `false` if the `configValue` is falsy? So you probably need undefined check at minimum?

matthew-chirgwin replied:
Good spot! Happy for this to land in next PR?

nictownsend commented on config/configHelpers.ts:
```diff
@@ -0,0 +1,75 @@
+          typeof configValue === 'function' ? configValue() : configValue;
+        publicValue = publicConfigValue;
+      } else {
+        const { values, publicValues } = processConfig([value]);
```
Being a generic function - what happens if you supply a `D` that has children that don't match `programmaticValue`, `literalValue` or `D`? 

I'm (guessing) typescript will complain - but it means for e.g that if you passed in `{type: ["x"]}` it should either infinite loop or blow up.

nictownsend replied:
~It also seems to suggest that every returned config value is of type `T` (so I'm guessing you then just have to pass in `T` as `unknown` or a subset.~

Saw later down how you can pass it through. I think the generic does work, but maybe the comment needs a bit more clarity to suggest that every value in `D` needs to be of type `T`

nictownsend replied:
It almost suggests that `D` needs `T` as a parameter somehow too (like `<D, T>(config: Array<D<T>>) => configurationType<T>`)

matthew-chirgwin replied:
> Being a generic function - what happens if you supply a `D` that has children that don't match `programmaticValue`, `literalValue` or `D`?
> 
> I'm (guessing) typescript will complain - but it means for e.g that if you passed in `{type: ["x"]}` it should either infinite loop or blow up.

It will - and with the config we have, the code wont compile without it being the right type (without config changes).

Will look to update typing now - when used I do specify the exact D and T https://github.com/matthew-chirgwin/strimzi-ui/blob/f3feb7ea63fabb60d9ee7e47d11de13e1b25f5f4/config/index.ts#L18 - but agree reading the function in isolation will make this hard to follow

nictownsend commented on server/README.md:
```diff
@@ -14,3 +14,27 @@ This directory contains all server code for the Strimzi UI - ie code which is re
+
+| Configuration                | Required | Default                                                                                                                | Purpose                                                                                                                                                                                |
+| ---------------------------- | -------- | ---------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| authentication.strategy      | No       | `none`                                                                                                                 | What authentication strategy to use to authenticate users. See [the security section](#security) for details of the available options.                                                 |
```
Overlap with security PR but this needs `authorization.strategy/configuration` too

matthew-chirgwin replied:
👍 - Would this come as a part of the security work? Main motivation with the movement of the config table from architecture to here was to slim down the architecture doc/keep the config with the code it acts on

nictownsend replied:
Yep that's fine :)

nictownsend commented on server/core/app.ts:
```diff
@@ -14,9 +14,8 @@ import {
-  serverName: string,
   getConfig: () => serverConfig
-) => express.Application = (serverName, getConfig) => {
+) => express.Application = (getConfig) => {
```
Why no server name?

matthew-chirgwin replied:
It isnt needed! As there is another parameter the linter didnt complain, but this was never used

nictownsend commented on tsconfig.base.json:
```diff
@@ -12,7 +12,7 @@
+    "outDir": "./js"
   },
-  "exclude": ["../../utils"]
+  "exclude": ["utils", "build", "**/test_common", "**/*.js"]
```
As this is being extended by `/client` and `/server` - are the relative paths not needed? (Including `outdir`)

And why are we ignoring `.js` files? I thought we were using `allowJS` to ensure we transpile the compiled JS + any source JS to the right es module support?

matthew-chirgwin replied:
At one stage I ran into an issue with the path config, show did a showConfig tsc run to debug. There I saw all utility, build, test code and linter (symlinks) + jest config being referenced in both the client and server outputs. I don't know how, but it seems when extended, the exclude is witth respect to where the config that defines it is. Hence the traversal didnt work.

This is an additional change - one I can back out. But with it, The config looked as expected for the client and server

nictownsend replied:
I trust you :)

nictownsend commented on server/test_common/testGQLRequests.ts:
```diff
@@ -0,0 +1,21 @@
+ * Mock requests used to test the gql elements of the server. Query names referenced in feature files, E.g `When I make a 'mockTopicsQuery' gql request to ...`. These can be generated using GraphiQL and it' export feature
+ */
+const mockTopicsQuery = {
+  query: '{\n  topics {\n    name\n    partitions\n    replicas\n  }\n}\n',
```
Does this need to be a string? Can supertest take an object as input?

matthew-chirgwin replied:
`These can be generated using GraphiQL and it' export feature`

Suspect using graphql tag it could be a typical query definition

Will fix the it' in the next PR

matthew-chirgwin commented:
For completeness, I have addressed @nictownsend comments in #99  - which once this is merged, will be be converted to 'full' PR

JQrdan commented on client/Contexts/ConfigFeatureFlag/ConfigFeatureFlag.spec.tsx:
```diff
@@ -0,0 +1,59 @@
+  describe('initial state tests', () => {
+    it('returns the expected configuration values', () => {
+      const { getByText } = render(
+        <ConfigFeatureFlagProvider>
```
Did you find this was easier than using the provider helper?

matthew-chirgwin replied:
The `ConfigFeatureFlagProvider` isnt a regular provider - it is a component which makes a query that provides a value - thus I had to drive it as such. As you will see in #99 though, I do use your helper for anything which consumes only :)

JQrdan reviewed (approved):
LGTM, nothing to add ontop of Nic's review

nictownsend committed:
feat: Config type rearrange (#6)

- refactor to enforcing typing for configs

Contributes to: strimzi/strimzi-ui#13

Signed-off-by: nictownsend <nictownsend@uk.ibm.com>

matthew-chirgwin commented:
Any thoughts @pmuir / @dlabaj ?
